torch.load default encoding change to 'utf-8'

nmilosev opened the pull request:
Default encoding when using torch.load to 'utf-8'
    
This commit provides changes for cases where user tries to torch.load 
a pickled module with non-ASCII characters in the docstring as 
discussed in #21743. The default encoding was changed from 'ascii'
to 'utf-8'. Documentation for `torch.load` was updated and two tests
(loading py2 unicode module with unicode in it; error throwing when
user explicitly sets wrong encoding) were written.


~~This commit provides changes for better error handling in cases
where user tries to `torch.load` a pickled module with non-ASCII
characters in the docstring as discussed in #21743.~~

Ping @ezyang

nmilosev commented:
One test needs to be modified further, I need to to remove the blank line from tests, and rebase. Will submit again shortly @ezyang sorry.

nmilosev commented:
Just pushed again, squashed to one commit, hope that's ok.

nmilosev commented:
@ezyang can you advise please? My test is failing on py2 because it doesn't raise the expected exception. Is it ok to write a py3 test only? That's the only thing left before everything is ok, I hope. Thanks! :)

ezyang commented:
> My test is failing on py2 because it doesn't raise the expected exception. Is it ok to write a py3 test only?

Yes. Indeed, that is exactly what the error is about; it doesn't occur if you go Py2 to Py2.

nmilosev commented:
@ezyang any clues on what is going on here?

```
16:29:10 Finished Build : #67597 of Job : caffe2-builds/py2-cuda9.0-cudnn7-windows-build with status : SUCCESS
16:29:10 FATAL: Channel "unknown": Remote call on JNLP4-connect connection from 54.167.77.194/54.167.77.194:41078 failed. The channel is closing down or has closed down
16:29:10 java.nio.channels.ClosedChannelException
16:29:10 	at org.jenkinsci.remoting.protocol.NetworkLayer.onRecvClosed(NetworkLayer.java:154)
16:29:10 	at org.jenkinsci.remoting.protocol.impl.NIONetworkLayer.ready(NIONetworkLayer.java:142)
16:29:10 	at org.jenkinsci.remoting.protocol.IOHub$OnReady.run(IOHub.java:795)
16:29:10 	at jenkins.util.ContextResettingExecutorService$1.run(ContextResettingExecutorService.java:28)
16:29:10 	at jenkins.security.ImpersonatingExecutorService$1.run(ImpersonatingExecutorService.java:59)
16:29:10 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
16:29:10 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
16:29:10 	at java.lang.Thread.run(Thread.java:748)
16:29:10 Caused: hudson.remoting.ChannelClosedException: Channel "unknown": Remote call on JNLP4-connect connection from 54.167.77.194/54.167.77.194:41078 failed. The channel is closing down or has closed down
16:29:10 	at hudson.remoting.Channel.call(Channel.java:948)
16:29:10 	at hudson.Launcher$RemoteLauncher.kill(Launcher.java:1085)
16:29:10 	at hudson.model.AbstractBuild$AbstractBuildExecution.run(AbstractBuild.java:510)
16:29:10 	at com.tikal.jenkins.plugins.multijob.MultiJobBuild$MultiJobRunnerImpl.run(MultiJobBuild.java:148)
16:29:10 	at hudson.model.Run.execute(Run.java:1810)
16:29:10 	at com.tikal.jenkins.plugins.multijob.MultiJobBuild.run(MultiJobBuild.java:76)
16:29:10 	at hudson.model.ResourceController.execute(ResourceController.java:97)
16:29:10 	at hudson.model.Executor.run(Executor.java:429)
16:29:10 [BFA] Scanning build for known causes...
16:29:10 [BFA] No failure causes found
16:29:10 [BFA] Done. 0s
16:29:11 Unable to get pull request builder trigger!!
16:29:11 Setting status of 584ff262d28cbd62ba52ce267d635a56d4d4b4d3 to FAILURE with url https://ci.pytorch.org/jenkins/job/caffe2-builds/job/py2-cuda9.0-cudnn7-windows-trigger-build/70422/ and message: 'Build failed
16:29:11  '
16:29:11 Using context: pr/caffe2-py2-cuda9.0-cudnn7-windows-build
16:29:11 Finished: FAILURE
```

Says build is SUCCESS but because of network failures whole job got FAILURE?

ezyang commented:
CI flakiness, don't worry about it.

driazati commented on torch/serialization.py:
```diff
@@ -606,7 +608,14 @@ def persistent_load(saved_id):
     _sys_info = pickle_module.load(f, **pickle_load_args)
     unpickler = pickle_module.Unpickler(f, **pickle_load_args)
```
From discussions in #21743 it's probably fine to change the default encoding here to `utf-8` and remove the `try-catch` below since that shouldn't be an issue anymore

nmilosev replied:
Thank you for the review! My only suggestion is to keep the `try-except` block even though `utf-8` is default. Say the user specifies `ascii` encoding and they try to load a module with `utf-8` in the docstring, they would still get the weird error like now. So my proposition is to keep the try-except __and__ set the `utf-8` to be default. I know this is really an edge case and it's probably never going to happen, but it doesn't hurt to keep the error handling part. :) 

Let me know what you think and I will send the corrected code.

Thank you again for the help!

nmilosev replied:
@driazati any update on this?

ezyang replied:
I think it's OK if you want to preserve the try-catch, but you are going to have to be super careful about the wording of the error message. The current wording is just not applicable to cases when an error would be thrown! I think it might be easiest to just remove the try-catch and if a new error case comes up, we know what specific case to write the error message for.

nmilosev replied:
I understand. Going to remove it then and update the PR. Thanks!

driazati commented on test/test_torch.py:
```diff
@@ -5875,6 +5875,17 @@ def test_serialization_storage_slice(self):
         self.assertEqual(s2[0], 0)
         self.assertEqual(s1.data_ptr() + 4, s2.data_ptr())
 
+    def test_load_unicode_error_msg(self):
```
This test can (mostly) be kept around to make sure `legacy_conv2d.pt` loads without error.

driazati reviewed (commented):
Good start! As for the tests our CI can be flaky at times, so if the failure doesn't look related it's probably not your fault.

nmilosev commented on torch/serialization.py:
```diff
@@ -419,6 +421,8 @@ def load(f, map_location=None, pickle_module=pickle, **pickle_load_args):
         new_fd = True
         f = f.open('rb')
     try:
+        if 'encoding' not in pickle_load_args.keys():
```
@ezyang and @driazati Please advise if this way of specifying the default parameter is ok. I thought about adding `encoding` as a named parameter but that would require more than 10 changes to `unpickler.load` function call so this seemed like a more elegant solution.

nmilosev commented:
I think I caught another bug! Passing `encoding` to `torch.load` does not work on Python2!

In `test_scriptability_serialization`:

```
Sep 23 15:47:14             with open(os.path.join(tmpdir, 'storages'), 'rb', 0) as f:
Sep 23 15:47:14 >               num_storages = pickle_module.load(f, **pickle_load_args)
Sep 23 15:47:14 E               TypeError: load() takes no keyword arguments
```

and also in test_resnet and many other tests. So on Python 2 we can't pass encoding via `pickle_load_args`. 

Also here: https://github.com/lisa-lab/pylearn2/issues/1298 the same problem but the bug was never reported.

@ezyang I think the way to deal with this is: if python version is 2 and user provides `pickle_load_args`, throw an exception?

I will temporarily solve this issue by only defaulting to `utf-8` on Python 3, but we should probably mention that `pickle_load_args` is Python 3 only.

Let me know what you think! :)

ezyang commented on torch/serialization.py:
```diff
@@ -391,8 +391,8 @@ def load(f, map_location=None, pickle_module=pickle, **pickle_load_args):
-        to strings using ``latin1`` encoding, and :attr:`encoding='bytes'` keeps them
+        these objects should be loaded, e.g., :attr:`encoding='utf-8'` decodes them
+        to strings using ``utf-8`` encoding, and :attr:`encoding='bytes'` keeps them
         as byte arrays which can be decoded later with ``byte_array.decode(...)``.
```
Probably want to mention we default to utf-8 here?

nmilosev replied:
Yes, good idea. Will add in a bit.

ezyang reviewed (approved):
lgtm

nmilosev committed:
Default encoding when using torch.load to 'utf-8' (py3)

This commit provides changes for cases where user tries to torch.load
a pickled module with non-ASCII characters in the docstring as
discussed in #21743. The default encoding was changed from 'ascii'
to 'utf-8'. Documentation for `torch.load` was updated and two tests
(loading py2 unicode module with unicode in it; error throwing when
user explicitly sets wrong encoding) were written.

Comment about `pickle_load_args` being unsupported on Python 2 was
also added. Also added a note about the new default.

nmilosev commented:
Just updated that last note, let me know what are the next steps. :)

PS. Thanks @ezyang for guiding me through everything!

ezyang committed:
Slight wording updates

ezyang commented:
@pytorchbot rebase this please

Commit:
Merge remote-tracking branch 'origin/master' into HEAD
